backend/src/main/kotlin/dres/run/SynchronousRunManager.kt

```diff
@@ -79,63 +92,81 @@ class SynchronousRunManager(competitionDescription: CompetitionDescription, name
             this.run.currentTask?.data?.submissions ?: emptyList()
         }
 
+    /** The list of all [Submission]s tracked ever received by this [SynchronousRunManager]. */
     override val allSubmissions: List<Submission>
         get() = this.stateLock.read {
             this.run.runs.flatMap { it.data.submissions }
         }
 
+    /** Internal data structure that tracks all WebSocket connections and their ready state (for [RunManagerStatus.PREPARING_TASK]) */
+    private val readyLatch = ReadyLatch<String>()
+
+    /** The internal [ScoreboardsUpdatable] instance for this [SynchronousRunManager]. */
+    override val scoreboards = ScoreboardsUpdatable(this.competitionDescription.generateDefaultScoreboards(), this.run)
+
+    /** The internal [MessageQueueUpdatable] instance used by this [SynchronousRunManager]. */
+    private val messageQueue = MessageQueueUpdatable(RunExecutor)
+
+    /** The internal [DAOUpdatable] instance used by this [SynchronousRunManager]. */
+    private val daoUpdatable = DAOUpdatable(RunExecutor.runs, this.run)
+
+    /** List of [Updatable] held by this [SynchronousRunManager]. */
+    private val updatables = mutableListOf<Updatable>()
+
+    /** Lock for changes to the [Updatable] list. */
+    private val updatableLock = StampedLock()
 
     /** The pipeline for [Submission] processing. All [Submission]s undergo three steps: filter, validation and score update. */
     private val submissionPipeline: List<Triple<SubmissionFilter,SubmissionValidator, TaskRunScorer>> = LinkedList()
 
     /** A lock for state changes to this [SynchronousRunManager]. */
     private val stateLock = ReentrantReadWriteLock()
 
-    /** Internal counter used to count the number of WebSocket connected clients. */
-    @Volatile
-    private var clientCounter = 0
-
-    /**
-     * Internal counter used to count the number ACKs received since the last issuing of a
-     * [ServerMessageType.TASK_PREPARE] message. -1 in cases where no [Task] is running.
-     */
-    @Volatile
-    private var ackCounter = -1
-
-    /** A flag indicating whether the [Scoreboard]s need an update. */
-    @Volatile
-    private var scoreboardUpdateRequired = true
-
     init {
-        this.run.id = this.dao.append(this.run)
-        //lambda to send submission updates
-        this.run.updateSubmissionStatus = {
-            this.executor.broadcastWsMessage(ServerMessage(this.runId, ServerMessageType.TASK_UPDATED))
-            scoreboardUpdateRequired = true
-        }
+        /* IMPORTANT: Order matters! */
+        this.updatables.add(this.scoreboards) /* 1: Update scoreboards. */
+        this.updatables.add(this.messageQueue) /* 2: Send WebSocket messages. */
+        this.updatables.add(this.daoUpdatable) /* 3: Update changes to the CompetitionRun object. */
```

Comment:
If the order is important, it should probably be encoded into the updatables themselves, so new ones can be added afterwards while still maintaining the relevant dependencies. On every change of the list, it would then simply need to be sorted by the priority.

Reply:
Interesting idea! I formalised this into the notion of a **Phase** or three of them to be more precise.